### ISSUE TITLE
Questions on performing PGD on the loss line from PID.py

### ISSUE THREAD

**Issue by nietzsche9088** [PKU-ML/Diffusion-PID-Protection#1]
The frequently used PGD is like "delta += alpha*Loss.grad().sgn() ", but  yours is "delta += delta.grad().sgn()" from line 258. I'm curious about why not compute the gradient on loss, but on delta. Thanks in advance.

**Comment by Charles20021201** [comment]
Hi @nietzsche9088:

Thanks for your attention to our work. I shall explain the implementation as follows:

As we all know, the PGD algorithm tries to solve $$\delta^* = \arg\max_{\delta\in [-\epsilon, \epsilon]} L( x + \delta)$$ by iteratively updating $\delta$ with $\delta^\prime = \text{Proj} ( \delta + \text{Sign}(\nabla_\delta L(x + \delta)) )$.

In our implementation, we first calculate the gradient of $\delta$ w.r.t. the loss, using `loss.backward().` Then we update the $\delta$ with `delta += delta.grad.sgn()`. And finally we project the $\delta$ into the $\epsilon$ ball using $torch.clip()$.

Our implementation follows the official implementation in the torchattcaks library and I kindly refer you to their implementation via this link. (https://adversarial-attacks-pytorch.readthedocs.io/en/latest/_modules/torchattacks/attacks/pgd.html#PGD). 

If you have further concerns, please let us know!

Best,
Charles

**Comment by nietzsche9088** [comment]
Hi Charles, 

Thanks for your reply. 

I've looked through the official implementation of PGD you shared.  They update the adversarial images by this line 
**grad = torch.autograd.grad(cost, adv_images, retain_graph=False, create_graph=False)[0]** . They get the gradient of cost (loss) on **adv_img**.  In your case, you calculate the gradient of \delta w.r.t. the loss. Is there an assumption that the gradient of \delta w.r.t. loss equals to the gradient of loss w.r.t. \delta?

Thanks,
Qiuyu


> Hi @nietzsche9088:
> 
> Thanks for your attention to our work. I shall explain the implementation as follows:
> 
> As we all know, the PGD algorithm tries to solve δ ∗ = arg ⁡ max δ ∈ [ − ϵ , ϵ ] L ( x + δ ) by iteratively updating δ with δ ′ = Proj ( δ + Sign ( ∇ δ L ( x + δ ) ) ) .
> 
> In our implementation, we first calculate the gradient of δ w.r.t. the loss, using `loss.backward().` Then we update the δ with `delta += delta.grad.sgn()`. And finally we project the δ into the ϵ ball using t o r c h . c l i p ( ) .
> 
> Our implementation follows the official implementation in the torchattcaks library and I kindly refer you to their implementation via this link. (https://adversarial-attacks-pytorch.readthedocs.io/en/latest/_modules/torchattacks/attacks/pgd.html#PGD).
> 
> If you have further concerns, please let us know!
> 
> Best, Charles

**Comment by Charles20021201** [comment]
> Hi Charles,
> 
> Thanks for your reply.
> 
> I've looked through the official implementation of PGD you shared. They update the adversarial images by this line **grad = torch.autograd.grad(cost, adv_images, retain_graph=False, create_graph=False)[0]** . They get the gradient of cost (loss) on **adv_img**. In your case, you calculate the gradient of \delta w.r.t. the loss. Is there an assumption that the gradient of \delta w.r.t. loss equals to the gradient of loss w.r.t. \delta?
> 
> Thanks, Qiuyu
> 
> > Hi @nietzsche9088:
> > Thanks for your attention to our work. I shall explain the implementation as follows:
> > As we all know, the PGD algorithm tries to solve δ ∗ = arg ⁡ max δ ∈ [ − ϵ , ϵ ] L ( x + δ ) by iteratively updating δ with δ ′ = Proj ( δ + Sign ( ∇ δ L ( x + δ ) ) ) .
> > In our implementation, we first calculate the gradient of δ w.r.t. the loss, using `loss.backward().` Then we update the δ with `delta += delta.grad.sgn()`. And finally we project the δ into the ϵ ball using t o r c h . c l i p ( ) .
> > Our implementation follows the official implementation in the torchattcaks library and I kindly refer you to their implementation via this link. (https://adversarial-attacks-pytorch.readthedocs.io/en/latest/_modules/torchattacks/attacks/pgd.html#PGD).
> > If you have further concerns, please let us know!
> > Best, Charles

HI @nietzsche9088:

We first clarify that we refer to $\frac{\partial L}{\partial \delta}$ when we spoke about the gradient of delta in the first reply. As $\frac{\partial \delta}{\partial L}$ is obviously wrong and will definitely incur an error in auto-differentiation.

To make it clear, our implementation can be divided into four steps:

(1) Line 224-225: Forwarding the perturbed images and setting up the computation graph. Note that we enable `self.delta[index].requires_grad_(True)` in this stage.

(2) Line 229-249: Calculating the loss according to `args.attack_type`. This will create a scalar output node in the computation graph with which we can compute the gradient.

(3) Line 252-253: Clearing up previous calculations and doing backpropagation.

(4) Line 256-261: Updating $\delta$ with `\delta.grad.sign()`, a.k.a, $\text{Sign}(\frac{\partial L}{\partial \delta})$. Abd do necessary projections.

I hope our further explanations can address your confusion. 

Best,
Charles

**Comment by nietzsche9088** [comment]
Thanks for your clarifications.

**Comment by Charles20021201** [comment]
> Thanks for your clarifications.

You are welcome. If there are other concerns, please let us know.
